core/shared/src/main/scala-2.11-2.12/ChunkBuilder.scala

```diff
@@ -79,28 +81,52 @@ private[zio] object ChunkBuilder {
    * values.
    */
   final class Boolean extends ChunkBuilder[SBoolean] { self =>
-    private val arrayBuilder: ArrayBuilder[SBoolean] = {
-      new ArrayBuilder.ofBoolean
+
+    private val bytes: ArrayBuffer[SByte] = ArrayBuffer()
+    private var maxBitIndex: SInt         = 0
+
+    private def appendBit(b: SBoolean): Unit = {
+      if (b) {
+        if (maxBitIndex % 8 == 0) {
+          bytes += (1 << 7).toByte
+        } else {
+          val idx         = 7 - (maxBitIndex % 8)
+          val byte: SByte = (bytes.last | (1 << idx)).toByte
+          bytes.update(bytes.size - 1, byte)
+        }
+      } else {
+        if (maxBitIndex % 8 == 0) {
+          bytes += 0.toByte
+        }
+      }
+      maxBitIndex = maxBitIndex + 1
     }
-    def +=(a: SBoolean): this.type = {
-      arrayBuilder += a
+    def +=(b: SBoolean): this.type = {
+      appendBit(b)
       this
     }
-    def clear(): Unit =
-      arrayBuilder.clear()
-    def result(): Chunk[SBoolean] =
-      Chunk.fromArray(arrayBuilder.result())
+    def clear(): Unit = {
+      bytes.clear()
+      maxBitIndex = 0
+    }
+    def result(): Chunk[SBoolean] = {
+      val chunk = Chunk.fromArray(bytes.toArray)
+      BitChunk(chunk, 0, maxBitIndex)
+    }
     override def ++=(as: TraversableOnce[SBoolean]): this.type = {
```

Comment:
I don't think this implementation is safe because we only know that it is safe to traverse `as` once but we are traversing it twice right now (once for `as.size` and once for `as.foreach`). We may just need to delegate to doing `+=` with each `Boolean` here. Unfortunately if we don't have the ability to traverse more than once there isn't a whole lot else we can do than process the elements one at a time unless we want to materialize them into their own collection which is kind of repeating the world of the `ChunkBuilder`.